src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -2544,9 +2544,17 @@ void MacroAssembler::debug64(char* msg, int64_t pc, int64_t regs[])
   }
 }
 
+RegSet MacroAssembler::call_clobbered_registers() {
+  RegSet regs = RegSet::range(r0, r17) - RegSet::of(rscratch1, rscratch2);
+#ifndef R18_RESERVED
+  regs += r18_reserved;
+#endif
+  return regs;
+}
+
 void MacroAssembler::push_call_clobbered_registers() {
   int step = 4 * wordSize;
-  push(RegSet::range(r0, r18) - RegSet::of(rscratch1, rscratch2), sp);
+  push(call_clobbered_registers() - RegSet::of(rscratch1, rscratch2), sp);
```

Comment:
rscratch1 and rscratch2 definitely are call clobbered.

Reply:
It's not exactly wrong. The readability of the code seems to have suffered in translation.

Reply:
Looking at the actual change make to development head, this is a misapplied patch. To match head, this should be:

```
RegSet MacroAssembler::call_clobbered_registers() {
  RegSet regs = RegSet::range(r0, r17) - RegSet::of(rscratch1, rscratch2);
#ifndef R18_RESERVED
  regs += r18_tls;
#endif
  return regs;
}
void MacroAssembler::push_call_clobbered_registers() {
  int step = 4 * wordSize;
  push(call_clobbered_registers(), sp);
```
Strictly speaking it should also push rscratch1 and rscratch2 - they are call clobbered - but we don't care because any macro can trash rscratch1 and rscratch2, so there's no point saving and restoring them.

Reply:
Thanks for catching this @VladimirKempik.

Reply:
@theRealAph you're right about `push_call_clobbered_registers()`, but I think `call_clobbered_registers()` is actually fine the way it is. The register named `r18_tls` on tip has been renamed to `r18_reserved` in this backport.

Reply:
Looks to me like the only real issue here is `warn()`. `stop()` doesn't restore the registers, so that's fine. We do want to see all the registers in a register dump. It would be better if `pusha()` were left as it is, but `popa()` did not clobber r18.

Reply:
> The register named `r18_tls` on tip has been renamed to `r18_reserved` in this backport.

Why has it been renamed `r18_reserved` ?